File: apps/android/app/src/main/kotlin/ai/parkshield/android/feature/auth/data/RestAuthenticationRepository.kt
package ai.parkshield.android.feature.auth.data

import ai.parkshield.android.core.network.ApiClient
import ai.parkshield.android.core.network.ApiRequest
import ai.parkshield.android.core.network.ApiResult
import ai.parkshield.android.feature.auth.domain.AuthFailure
import ai.parkshield.android.feature.auth.domain.AuthResult
import ai.parkshield.android.feature.auth.domain.AuthenticatedUser
import ai.parkshield.android.feature.auth.domain.AuthenticationRepository
import ai.parkshield.android.feature.auth.domain.TokenPair
import org.json.JSONObject

class RestAuthenticationRepository(private val client: ApiClient) : AuthenticationRepository {
    override suspend fun login(email: String, password: CharArray): AuthResult<TokenPair> = try {
        executePair("/api/v1/auth/login", JSONObject().put("email", email).put("password", password.concatToString()))
    } finally {
        password.fill('\u0000')
    }

    override suspend fun refresh(refreshToken: String): AuthResult<TokenPair> =
        executePair("/api/v1/auth/refresh", JSONObject().put("refresh_token", refreshToken))

    override suspend fun logout(refreshToken: String): AuthResult<Unit> = when (val result = client.execute(
        ApiRequest("POST", "/api/v1/auth/logout", JSONObject().put("refresh_token", refreshToken).toString().encodeToByteArray()),
    )) {
        is ApiResult.Success -> if (result.statusCode == 204) AuthResult.Success(Unit) else AuthResult.Failure(AuthFailure.Transport)
        is ApiResult.Failure -> AuthResult.Failure(result.failure())
    }

    override suspend fun profile(accessToken: String): AuthResult<AuthenticatedUser> = when (val result = client.execute(
        ApiRequest("GET", "/api/v1/auth/me", headers = mapOf("Authorization" to "Bearer $accessToken")),
    )) {
        is ApiResult.Success -> try {
            val body = JSONObject(result.body.decodeToString())
            AuthResult.Success(AuthenticatedUser(body.getString("id"), body.getString("email"), body.getString("role"), body.getBoolean("is_verified")))
        } catch (_: Exception) { AuthResult.Failure(AuthFailure.Transport) }
        is ApiResult.Failure -> AuthResult.Failure(result.failure())
    }

    private suspend fun executePair(path: String, body: JSONObject): AuthResult<TokenPair> = when (val result = client.execute(
        ApiRequest("POST", path, body.toString().encodeToByteArray()),
    )) {
        is ApiResult.Success -> try {
            val json = JSONObject(result.body.decodeToString())
            AuthResult.Success(TokenPair(json.getString("access_token"), json.getString("refresh_token"), json.getLong("expires_in"), json.optString("token_type", "bearer")))
        } catch (_: Exception) { AuthResult.Failure(AuthFailure.Transport) }
        is ApiResult.Failure -> AuthResult.Failure(result.failure())
    }

    private fun ApiResult.Failure.failure(): AuthFailure = when (error?.code) {
        "SESSION_INVALID" -> AuthFailure.SessionInvalid
        null -> AuthFailure.Transport
        else -> AuthFailure.Remote(error.code, error.correlationId)
    }
}

File: apps/ios/Sources/ParkShieldFoundation/Feature/Auth/Data/RestAuthenticationRepository.swift
import Foundation

struct RestAuthenticationRepository: AuthenticationRepository {
    let client: any APIClient
    func login(email: String, password: String) async -> Result<TokenPair, AuthFailure> { await pair("/api/v1/auth/login", ["email": email, "password": password]) }
    func refresh(refreshToken: String) async -> Result<TokenPair, AuthFailure> { await pair("/api/v1/auth/refresh", ["refresh_token": refreshToken]) }
    func logout(refreshToken: String) async -> Result<Void, AuthFailure> {
        await execute(APIRequest(method: "POST", path: "/api/v1/auth/logout", body: json(["refresh_token": refreshToken]))) { result in
            if case .success(let status, _) = result, status == 204 { return .success(()) }; return .failure(self.failure(result))
        }
    }
    func profile(accessToken: String) async -> Result<AuthenticatedUser, AuthFailure> {
        await execute(APIRequest(method: "GET", path: "/api/v1/auth/me", headers: ["Authorization": "Bearer \(accessToken)"])) { result in
            guard case .success(_, let data) = result, let user = try? JSONDecoder().decode(AuthenticatedUser.self, from: data) else { return .failure(self.failure(result)) }; return .success(user)
        }
    }
    private func pair(_ path: String, _ payload: [String: String]) async -> Result<TokenPair, AuthFailure> {
        await execute(APIRequest(method: "POST", path: path, body: json(payload))) { result in
            guard case .success(_, let data) = result, let pair = try? JSONDecoder().decode(TokenPair.self, from: data) else { return .failure(self.failure(result)) }; return .success(pair)
        }
    }
    private func execute<T>(_ request: APIRequest, map: (APIResult) -> Result<T, AuthFailure>) async -> Result<T, AuthFailure> { do { return map(try await client.execute(request)) } catch { return .failure(.transport) } }
    private func json(_ value: [String: String]) -> Data? { try? JSONEncoder().encode(value) }
    private func failure(_ result: APIResult) -> AuthFailure { if case .failure(_, let error) = result { if error?.code == "SESSION_INVALID" { return .sessionInvalid }; if let error { return .remote(code: error.code, correlationID: error.correlationID) } }; return .transport }
}
